js/driver/virtio/blk.js

```diff
@@ -0,0 +1,141 @@
+// Copyright 2016-present runtime.js project authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+'use strict';
+const VirtioDevice = require('./device');
+const runtime = require('../../core');
+const { Uint64LE } = require('int64-buffer');
+
+const VIRTIO_BLK_T_IN = 0;
+const VIRTIO_BLK_T_OUT = 1;
+
+function initializeBlockDevice(pciDevice) {
+  const ioSpace = pciDevice.getBAR(0).resource;
+  const irq = pciDevice.getIRQ();
+
+  const features = {
+    VIRTIO_BLK_F_SIZE_MAX: 1,
+    VIRTIO_BLK_F_SEG_MAX: 2,
+    VIRTIO_BLK_F_GEOMETRY: 4,
+    VIRTIO_BLK_F_RO: 5,
+    VIRTIO_BLK_F_BLK_SIZE: 6,
+    VIRTIO_BLK_F_FLUSH: 9,
+    VIRTIO_BLK_F_TOPOLOGY: 10,
+    VIRTIO_BLK_F_CONFIG_WCE: 11,
+  };
+
+  const dev = new VirtioDevice('blk', ioSpace);
+  dev.setDriverAck();
+
+  const driverFeatures = {
+    // VIRTIO_BLK_F_SIZE_MAX: true,
+    VIRTIO_BLK_F_SEG_MAX: true,
+    VIRTIO_BLK_F_GEOMETRY: true,
+    VIRTIO_BLK_F_BLK_SIZE: true,
+    VIRTIO_BLK_F_TOPOLOGY: true,
+  };
+
+  const deviceFeatures = dev.readDeviceFeatures(features);
+  debug(JSON.stringify(deviceFeatures));
+
+  if (!dev.writeGuestFeatures(features, driverFeatures, deviceFeatures)) {
+    debug('[virtio] blk driver is unable to start');
+    return;
+  }
+
+  const QUEUE_ID_REQ = 0;
+
+  const reqQueue = dev.queueSetup(QUEUE_ID_REQ);
+  const promiseQueue = [];
+
+  const sectorSize = 512;
+  const sectorCount = dev.blkReadSectorCount();
+  const totalSectorCount = dev.blkReadTotalSectorCount();
+
+  function buildHeader(type, sector) {
+    const u8 = new Uint8Array(16);
+    const view = new DataView(u8.buffer);
+    view.setUint32(0, type, true);
+    view.setUint32(4, 0, true); // priority: low
+    u8.set((new Uint64LE(sector)).toArray(), 8);
+    return u8;
+  }
+
+  const diskDriver = new runtime.disk.DiskInterface('hda', {
```

Comment:
> good idea, js array disks[0] is nice, though not sure it should be
separated by buses, since all block devices share the same interface.
Easier to iterate and filter a plain array. Alternatively bus (and possibly
slot) can be a separate property on the device object.

I'm asking about a way to identify uniquely each device, and bus+address
seems to be the easiest way and the one more close to the actual
configuration, but later there would be other names more human-friendly
pointing to the same object. Runtime.js is a kernel, think low-level and
machine and performant-friendly first and left upper layers to make it
easier to use :-)